Content.Client/GameObjects/EntitySystems/DragDropSystem.cs

```diff
@@ -0,0 +1,402 @@
+
+            // remove current highlights
+            RemoveHighlights();
+
+            // find possible targets on screen even if not reachable
+            // TODO: Duplicated in SpriteSystem
+            var pvsBounds = _eyeManager.GetWorldViewport().Enlarged(5);
+            var pvsEntities = EntityManager.GetEntitiesIntersecting(_eyeManager.CurrentMap, pvsBounds, true);
+            foreach (var pvsEntity in pvsEntities)
+            {
+                if (pvsEntity.TryGetComponent<SpriteComponent>(out var inRangeSprite))
+                {
+                    // can't highlight if there's no sprite or it's not visible
+                    if (inRangeSprite.Visible == false) continue;
+
+                    // check if it's able to be dropped on by current dragged entity
+                    if (_draggable.ClientCanDropOn(new CanDropEventArgs(_dragger, _draggedEntity, pvsEntity)))
+                    {
+                        // highlight depending on whether its in or out of range
+                        var inRange = _interactionSystem.InRangeUnobstructed(_dragger.Transform.MapPosition,
+                            pvsEntity.Transform.MapPosition, ignoredEnt: _dragger);
+                        inRangeSprite.PostShader = inRange ? _dropTargetInRangeShader : _dropTargetOutOfRangeShader;
+                        inRangeSprite.RenderOrder = EntityManager.CurrentTick.Value;
+                        highlightedSprites.Add(inRangeSprite);
+                    }
+                }
+            }
+        }
+
+        private void RemoveHighlights()
+        {
+            foreach (var highlightedSprite in highlightedSprites)
+            {
+                highlightedSprite.PostShader = null;
+                highlightedSprite.RenderOrder = 0;
+            }
+            highlightedSprites.Clear();
+        }
+
+        /// <summary>
+        /// Cancels the drag, firing our saved drag event if instructed to do so and
+        /// we are within the threshold for replaying the click
+        /// (essentially reverting the drag attempt and allowing the original click
+        /// to proceed as if no drag was performed)
+        /// </summary>
+        private void CancelDrag(bool fireSavedCmd)
+        {
+            RemoveHighlights();
+            if (_dragShadow != null)
+            {
+                _entityManager.DeleteEntity(_dragShadow);
+            }
+
+            _dragShadow = null;
+            _draggedEntity = null;
+            _draggable = null;
+            _dragger = null;
+            _state = DragState.NOT_DRAGGING;
+
+            _mouseDownTime = 0;
+
+            if (fireSavedCmd && _savedMouseDown.HasValue && _mouseDownTime < MaxMouseDownTimeForReplayingClick)
+            {
+                var savedValue = _savedMouseDown.Value;
+                _isReplaying = true;
+                _inputSystem.HandleInputCommand(savedValue.Session, EngineKeyFunctions.Use,
```

Comment:
Ok so the replay thing is probably necessary, although you probably *should* go and change it so that the new input message sent has timing info based on the current tick instead of when you saved it.